tensornetwork/backends/numpy/numpy_backend.py

```diff
@@ -283,7 +283,146 @@ def matvec(vector):
+                 Default: None.
+      A_kwargs : Keyword arguments to `A_mv`, supplied to this interface
+                 as a dictionary.
+                 Default: None.
+      x0       : An optional guess solution. Zeros are used by default.
+                 If `x0` is supplied, its shape and dtype must match those of
+                 `b`, or an
+                 error will be thrown.
+                 Default: zeros.
+      tol, atol: Solution tolerance to achieve,
+                 norm(residual) <= max(tol*norm(b), atol).
+                 Default: tol=1E-05
+                          atol=tol
+      num_krylov_vectors
+               : Size of the Krylov space to build at each restart.
+                 Expense is cubic in this parameter. If supplied, it must be
+                 an integer in 0 < num_krylov_vectors <= b.size.
+                 Default: b.size.
+      maxiter  : The Krylov space will be repeatedly rebuilt up to this many
+                 times. Large values of this argument
+                 should be used only with caution, since especially for nearly
+                 symmetric matrices and small `num_krylov_vectors` convergence
+                 might well freeze at a value significantly larger than `tol`.
+                 Default: 1.
+      M        : Inverse of the preconditioner of A; see the docstring for
+                 `scipy.sparse.linalg.gmres`. This is only supported in the
+                 numpy backend. Supplying this argument to other backends will
+                 trigger NotImplementedError.
+                 Default: None.
+
+    Raises:
+      ValueError: -if `x0` is supplied but its shape differs from that of `b`.
+                  -if the ARPACK solver reports a breakdown (which usually
+                   indicates some kind of floating point issue).
+                  -if num_krylov_vectors is 0 or exceeds b.size.
+                  -if tol was negative.
+
+    Returns:
+      x       : The converged solution. It has the same shape as `b`.
+      info    : 0 if convergence was achieved, the number of restarts otherwise.
+    """
+
+    if x0 is not None and x0.shape != b.shape:
+      errstring = "If x0 is supplied, its shape, " + str(x0.shape) + ", must \n"
+      errstring += "match that of b, " + str(b.shape) + "."
+      raise ValueError(errstring)
+    if x0 is not None and x0.dtype != b.dtype:
+      errstring = "If x0 is supplied, its dtype, " + str(x0.dtype) + ", must \n"
+      errstring += "match that of b, " + str(b.dtype) + "."
+      raise ValueError(errstring)
+    if num_krylov_vectors is None:
+      num_krylov_vectors = b.size
+    if num_krylov_vectors <= 0 or num_krylov_vectors > b.size:
+      errstring = "num_krylov_vectors must be in "
+      errstring += "0 < num_krylov_vectors <= b.size.\n"
+      errstring += "num_krylov_vectors = " + str(num_krylov_vectors) + ".\n"
+      errstring += "b.size = " + str(b.size) + "."
+      raise ValueError(errstring)
```

Comment:
small hint: you can break lines with strings wihthout having to do +=, and 
with f-strings (supported in python 3.4 and later I think) you can use {} to 
insert expressions into the string
```python 
errstring = f"num_krylov_vectors must be in "
            f"0 < num_krylov_vectors <= b.size.\n"
            f"num_krylov_vectors = {num_krylov_vectors}.\n"
            f"b.size = {b.size}."
```

Reply:
Oh, neat!